frontend/types.tsx

```diff
@@ -160,3 +160,38 @@ export interface CoursePageProps {
     course: Course;
     leadership: Membership[];
 }
+
+export enum Metric {
+    HEATMAP_WAIT = "HEATMAP_AVG_WAIT",
+    HEATMAP_QUESTIONS = "HEATMAP_QUESTIONS_PER_TA",
+    AVG_WAIT = "AVG_WAIT",
+    NUM_ANSWERED = "NUM_ANSWERED",
+    STUDENTS_HELPED = "STUDENTS_HELPED",
+    AVG_TIME_HELPING = "AVG_TIME_HELPING",
+    LIST_WAIT = "LIST_WAIT_TIME_DAYS",
+}
+
+interface AnalyticsDataElem {
+    metric: Metric;
+    day: number;
```

Comment:
According to the django model, `day`, `hour`, and `date` could be undefined. Those fields may have to be optional here

Reply:
Ah good catch - some metrics have day/hour/date and some don't, but this is known ahead of time so @conswang can you turn this into a discriminated union with metric being the discriminator?

Reply:
Sure, so something like AnalyticsDataElem = HeatmapDataElem | AverageDataElem | ...  where each type may or may not have date/hour/etc, right

Also, I've just noticed that we seem to be returning all the fields from the backend for most? metrics, though they may be null (eg. `date` for heatmaps, `day` and `hour` for averages). Is it safe to assume that the fields that are null are the ones we don't need for a particular metric?

![image](https://user-images.githubusercontent.com/52753462/106415367-e9d75f80-641c-11eb-8895-fb745623748d.png)

Reply:
Yeap! And as for backend response, yes we can assume that for heatmaps there won't be a date because the values are averages, and for averages, they are day-to-day statistics that give you data to plot graphs, but don't have hour granularity.

Reply:
Fixed it I think! I actually didn't put the union type since we wouldn't really be using it for anything, we always want either an array of heatmap data or an array of another kind of averages data (with metric, value, and date)